Fix some bootstrap papercuts

WaffleLapkin opened the pull request:
... related to jj and my `./build` symlink setup[^1].

I'm not sure if these are good solutions, but they seem to work. See commits for a bit more info.

r? @jieyouxu 

[^1]: see #139804

WaffleLapkin committed:
update submodules if the directory doesn't exist

WaffleLapkin committed:
fix typo

WaffleLapkin committed:
canonicalize test build dir before normalizing it

Fix fixes failures of the following tests when build directory is a
symlink:
- `tests/ui/error-codes/E{0464,0523}.rs`
- `tests/ui/crate-loading/crateresolve{1,2}.rs` (those are the same tests)

jieyouxu commented on src/bootstrap/src/core/config/config.rs:
```diff
@@ -2888,6 +2888,13 @@ impl Config {
 
         let absolute_path = self.src.join(relative_path);
 
+        // NOTE: This check is required because `jj git clone` doesn't create directories for
+        // submodules, they are completely ignored. The code below assumes this directory exists,
+        // so create it here.
+        if !absolute_path.exists() {
+            t!(fs::create_dir_all(&absolute_path));
+        }
```
Remark: `Path::exists` will traverse symlinks to look for existence of the destination file (i.e. not necessarily the "shallow" `$checkout_root/$usual_submodule_location`), but I guess if the user has a symlink in the place of the usual submodules in checkout...

jieyouxu commented on src/tools/compiletest/src/runtest.rs:
```diff
@@ -2375,9 +2375,13 @@ impl<'test> TestCx<'test> {
         let rust_src_dir = rust_src_dir.read_link_utf8().unwrap_or(rust_src_dir.to_path_buf());
         normalize_path(&rust_src_dir.join("library"), "$SRC_DIR_REAL");
 
+        // Canonicalize test build directory path.
+        // Without this some tests fail if build directory is a symlink.
+        let output_base_dir = fs::canonicalize(self.output_base_dir()).unwrap();
```
Suggestion: I just switched compiletest over to use `camino` to enforce UTF-8 paths, so this should instead be

```rs
let output_base_dir = self.output_base_dir().canonicalize_utf8().unwrap();
```

jieyouxu replied:
Discussion: that being said, I'm not 100% sure this is sufficient in itself, since rustc itself might have inconsistencies when `build/` is a symlink to elsewhere...

WaffleLapkin replied:
(note that there are more `fs::canonicalize` in the file, not sure if they are intended)

The tests seem to pass for me, so this is *probably* fine. I'm more concerned about a case where the directory doesn't exist, or, windows where canonicalize might return weird paths:
> On Windows, this converts the path to use [extended length path](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file) syntax, which allows your program to use longer path names, but means you can only join backslash-delimited paths to it, and it may be incompatible with other applications (if passed to the application on the command-line, or written to a file another application may read).

jieyouxu replied:
Yeah, some of these runtest bits need further auditing, which is beyond the scope of this PR.

jieyouxu reviewed (approved):
Thanks

jieyouxu commented:
@bors r+ rollup

bors commented:
:pushpin: Commit f038953c932e3756f08bf62f8017797f4d9aab73 has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu f038953c932e3756f08bf62f8017797f4d9aab73 -->
<!-- homu: {"type":"Approved","sha":"f038953c932e3756f08bf62f8017797f4d9aab73","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

WaffleLapkin commented:
I ran more tests and some are broken (potentially as a result of this PR), so @bors r-

WaffleLapkin committed:
normalize canonical and non-canonical paths in compiletest

Apparently there are tests that print canonical paths *and* tests which
print non-canonical paths.

An example of the latter is `tests/ui/type_length_limit.rs`.

WaffleLapkin commented:
@rustbot review
(see new commit)

jieyouxu reviewed (approved):
Huh... I guess I'm not really surprised that this isn't super consistent, lol. Thanks

jieyouxu commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 89b4eba49cc15e0a4e53f7816473201d12be5f4f has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu 89b4eba49cc15e0a4e53f7816473201d12be5f4f -->
<!-- homu: {"type":"Approved","sha":"89b4eba49cc15e0a4e53f7816473201d12be5f4f","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
